File: android/app/src/main/java/com/cbtipul/app/data/DiaryOne.kt
package com.cbtipul.app.data

import com.cbtipul.app.model.DatabaseId
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

enum class DiaryOneEntryCreator(val raw: String) {
    Therapist("therapist"),
    Patient("patient"),
    ;

    companion object {
        fun fromRaw(value: String): DiaryOneEntryCreator =
            entries.find { it.raw == value } ?: Therapist
    }
}

data class DiaryOneEntry(
    val id: String,
    val patientId: DatabaseId,
    val therapistId: String,
    val createdBy: DiaryOneEntryCreator,
    val event: String,
    val thought: String,
    val feelings: List<DiaryFeeling>,
    val behaviour: String,
    val physicalSymptoms: String?,
    val createdAt: Date,
    val updatedAt: Date,
)

@Serializable
private data class DiaryOneEntryRow(
    val id: String,
    @SerialName("patient_id") val patientId: String,
    @SerialName("therapist_id") val therapistId: String,
    @SerialName("created_by") val createdBy: String,
    val event: String,
    val thought: String,
    val feelings: List<DiaryFeeling> = emptyList(),
    val behaviour: String,
    @SerialName("physical_symptoms") val physicalSymptoms: String? = null,
    @SerialName("created_at") val createdAt: String,
    @SerialName("updated_at") val updatedAt: String,
)

class DiaryOneRepository(private val client: SupabaseClient) {
    private val _entries = MutableStateFlow<Map<String, List<DiaryOneEntry>>>(emptyMap())
    val entries: StateFlow<Map<String, List<DiaryOneEntry>>> = _entries.asStateFlow()
    private val demoEntries = mutableMapOf<String, List<DiaryOneEntry>>()

    fun entriesFor(patientId: DatabaseId): List<DiaryOneEntry> =
        _entries.value[patientId.queryValue].orEmpty()

    suspend fun loadEntries(patientId: DatabaseId): List<DiaryOneEntry> {
        if (DemoData.isDemoId(patientId)) {
            val cached = demoEntries[patientId.queryValue].orEmpty()
            _entries.update { it + (patientId.queryValue to cached) }
            return cached
        }
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        val patientUuid = PatientAssignmentRepository.uuidOrNull(patientId)
            ?: throw IllegalStateException("not_configured")
        val rows = client.from("diary_one_entries")
            .select(columns) {
                filter { eq("patient_id", patientUuid) }
                order("created_at", Order.DESCENDING)
            }
            .decodeList<DiaryOneEntryRow>()
        val loaded = rows.map { it.toDomain(patientId) }
        _entries.update { it + (patientId.queryValue to loaded) }
        return loaded
    }

    suspend fun createEntry(
        patientId: DatabaseId,
        event: String,
        thought: String,
        feelings: List<DiaryFeeling>,
        behaviour: String,
        physicalSymptoms: String?,
    ): DiaryOneEntry {
        val symptoms = physicalSymptoms?.trim()?.ifEmpty { null }
        if (DemoData.isDemoId(patientId)) {
            val entry = DiaryOneEntry(
                id = UUID.randomUUID().toString(),
                patientId = patientId,
                therapistId = UUID.randomUUID().toString(),
                createdBy = DiaryOneEntryCreator.Therapist,
                event = event,
                thought = thought,
                feelings = feelings,
                behaviour = behaviour,
                physicalSymptoms = symptoms,
                createdAt = Date(),
                updatedAt = Date(),
            )
            upsert(entry)
            return entry
        }
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        val patientUuid = PatientAssignmentRepository.uuidOrNull(patientId)
            ?: throw IllegalStateException("not_configured")
        val therapistId = client.auth.currentSessionOrNull()?.user?.id
            ?: throw IllegalStateException("not_signed_in")
        val body = buildJsonObject {
            put("patient_id", patientUuid)
            put("therapist_id", therapistId)
            put("created_by", DiaryOneEntryCreator.Therapist.raw)
            put("event", event)
            put("thought", thought)
            put("feelings", feelingsJson(feelings))
            put("behaviour", behaviour)
            if (symptoms == null) put("physical_symptoms", JsonNull) else put("physical_symptoms", symptoms)
        }
        val saved = client.from("diary_one_entries")
            .insert(body) { select(columns) }
            .decodeSingle<DiaryOneEntryRow>()
            .toDomain(patientId)
        upsert(saved)
        return saved
    }

    suspend fun updateEntry(
        id: String,
        patientId: DatabaseId,
        event: String,
        thought: String,
        feelings: List<DiaryFeeling>,
        behaviour: String,
        physicalSymptoms: String?,
    ): DiaryOneEntry {
        val symptoms = physicalSymptoms?.trim()?.ifEmpty { null }
        if (DemoData.isDemoId(patientId)) {
            val existing = demoEntries[patientId.queryValue].orEmpty().first { it.id == id }
            val updated = existing.copy(
                event = event,
                thought = thought,
                feelings = feelings,
                behaviour = behaviour,
                physicalSymptoms = symptoms,
                updatedAt = Date(),
            )
            upsert(updated)
            return updated
        }
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        val body = buildJsonObject {
            put("event", event)
            put("thought", thought)
            put("feelings", feelingsJson(feelings))
            put("behaviour", behaviour)
            if (symptoms == null) put("physical_symptoms", JsonNull) else put("physical_symptoms", symptoms)
            put("updated_at", timestampNow())
        }
        val saved = client.from("diary_one_entries")
            .update(body) {
                filter { eq("id", id) }
                select(columns)
            }
            .decodeSingle<DiaryOneEntryRow>()
            .toDomain(patientId)
        upsert(saved)
        return saved
    }

    suspend fun deleteEntry(id: String, patientId: DatabaseId) {
        if (DemoData.isDemoId(patientId)) {
            demoEntries[patientId.queryValue] =
                demoEntries[patientId.queryValue].orEmpty().filterNot { it.id == id }
            _entries.update { it + (patientId.queryValue to demoEntries[patientId.queryValue].orEmpty()) }
            return
        }
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        val deleted = client.from("diary_one_entries")
            .delete {
                filter { eq("id", id) }
                select(Columns.raw("id"))
            }
            .decodeList<DeletedId>()
        if (deleted.isEmpty()) throw IllegalStateException("not_configured")
        _entries.update { map ->
            val next = map[patientId.queryValue].orEmpty().filterNot { it.id == id }
            map + (patientId.queryValue to next)
        }
    }

    fun clear() {
        _entries.value = emptyMap()
        demoEntries.clear()
    }

    private fun upsert(entry: DiaryOneEntry) {
        val key = entry.patientId.queryValue
        val next = (listOf(entry) + entriesFor(entry.patientId).filterNot { it.id == entry.id })
            .sortedByDescending { it.createdAt.time }
        if (DemoData.isDemoId(entry.patientId)) demoEntries[key] = next
        _entries.update { it + (key to next) }
    }

    private fun feelingsJson(feelings: List<DiaryFeeling>) = buildJsonArray {
        feelings.forEach { feeling ->
            add(
                buildJsonObject {
                    put("name", feeling.name)
                    put("intensity", feeling.intensity)
                },
            )
        }
    }

    private fun timestampNow(): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        return formatter.format(Date())
    }

    private fun DiaryOneEntryRow.toDomain(patientId: DatabaseId) = DiaryOneEntry(
        id = id,
        patientId = patientId,
        therapistId = therapistId,
        createdBy = DiaryOneEntryCreator.fromRaw(createdBy),
        event = event,
        thought = thought,
        feelings = feelings,
        behaviour = behaviour,
        physicalSymptoms = physicalSymptoms,
        createdAt = parseIso(createdAt),
        updatedAt = parseIso(updatedAt),
    )

    @Serializable
    private data class DeletedId(val id: String)

    companion object {
        private val columns = Columns.raw(
            "id, patient_id, therapist_id, created_by, event, thought, feelings, behaviour, physical_symptoms, created_at, updated_at",
        )

        private fun parseIso(raw: String): Date {
            listOf(
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US),
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX", Locale.US),
            ).forEach { it.parse(raw)?.let { date -> return date } }
            return Date()
        }
    }
}

File: ios/DiaryOne.swift
import Foundation
import OSLog
import Supabase

enum DiaryOneEntryCreator: String, Codable, Sendable {
    case therapist
    case patient
}

/// One Diary 1 record for a Patient. Not tied to a Session or assignment.
nonisolated struct DiaryOneEntry: Identifiable, Equatable, Sendable, Codable {
    let id: UUID
    let patientId: DatabaseID
    let therapistId: UUID
    let createdBy: DiaryOneEntryCreator
    var event: String
    var thought: String
    var feeling: String
    var feelingIntensity: Int
    var behaviour: String
    var physicalSymptoms: String?
    let createdAt: Date
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case patientId = "patient_id"
        case therapistId = "therapist_id"
        case createdBy = "created_by"
        case event
        case thought
        case feeling
        case feelingIntensity = "feeling_intensity"
        case behaviour
        case physicalSymptoms = "physical_symptoms"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

private struct NewDiaryOneEntry: Encodable {
    let patientId: UUID
    let therapistId: UUID
    let createdBy: DiaryOneEntryCreator
    let event: String
    let thought: String
    let feeling: String
    let feelingIntensity: Int
    let behaviour: String
    let physicalSymptoms: String?

    enum CodingKeys: String, CodingKey {
        case patientId = "patient_id"
        case therapistId = "therapist_id"
        case createdBy = "created_by"
        case event, thought, feeling, behaviour
        case feelingIntensity = "feeling_intensity"
        case physicalSymptoms = "physical_symptoms"
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(patientId, forKey: .patientId)
        try container.encode(therapistId, forKey: .therapistId)
        try container.encode(createdBy, forKey: .createdBy)
        try container.encode(event, forKey: .event)
        try container.encode(thought, forKey: .thought)
        try container.encode(feeling, forKey: .feeling)
        try container.encode(feelingIntensity, forKey: .feelingIntensity)
        try container.encode(behaviour, forKey: .behaviour)
        try container.encode(physicalSymptoms, forKey: .physicalSymptoms)
    }
}

private struct DiaryOneEntryClinicalUpdate: Encodable {
    let event: String
    let thought: String
    let feeling: String
    let feelingIntensity: Int
    let behaviour: String
    let physicalSymptoms: String?
    let updatedAt: String

    enum CodingKeys: String, CodingKey {
        case event, thought, feeling, behaviour
        case feelingIntensity = "feeling_intensity"
        case physicalSymptoms = "physical_symptoms"
        case updatedAt = "updated_at"
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(event, forKey: .event)
        try container.encode(thought, forKey: .thought)
        try container.encode(feeling, forKey: .feeling)
        try container.encode(feelingIntensity, forKey: .feelingIntensity)
        try container.encode(behaviour, forKey: .behaviour)
        try container.encode(physicalSymptoms, forKey: .physicalSymptoms)
        try container.encode(updatedAt, forKey: .updatedAt)
    }
}

private struct DeletedDiaryOneRow: Decodable {
    let id: UUID
}

/// Therapist CRUD for `public.diary_one_entries`. RLS is the authorization
/// boundary. Demo clinic IDs stay in memory because they are not UUIDs.
@Observable
@MainActor
final class DiaryOneStore {
    private let client: SupabaseClient
    private var entriesByPatient: [String: [DiaryOneEntry]] = [:]
    /// Demo clinic only — not used for real patients.
    private var demoEntriesByPatient: [String: [DiaryOneEntry]] = [:]

    init(client: SupabaseClient) {
        self.client = client
    }

    func entries(for patientId: DatabaseID) -> [DiaryOneEntry] {
        cached(for: patientId)
    }

    func loadEntries(for patientId: DatabaseID) async throws -> [DiaryOneEntry] {
        if DemoData.isDemoID(patientId) {
            return cached(for: patientId)
        }
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        guard let patientUUID = patientId.uuidValue else {
            throw AuthError.notConfigured
        }
        do {
            let rows: [DiaryOneEntry] = try await client.from("diary_one_entries")
                .select(
                    "id, patient_id, therapist_id, created_by, event, thought, feeling, feeling_intensity, behaviour, physical_symptoms, created_at, updated_at"
                )
                .eq("patient_id", value: patientUUID)
                .order("created_at", ascending: false)
                .execute()
                .value
            entriesByPatient[patientId.queryValue] = rows
            AppLog.store.info("Diary 1 entries loaded")
            return rows
        } catch {
            AppLog.store.error(
                "Diary 1 load failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    func createEntry(
        patientId: DatabaseID,
        event: String,
        thought: String,
        feeling: String,
        feelingIntensity: Int,
        behaviour: String,
        physicalSymptoms: String?
    ) async throws -> DiaryOneEntry {
        let symptoms = Self.nullIfEmpty(physicalSymptoms)
        if DemoData.isDemoID(patientId) {
            let entry = DiaryOneEntry(
                id: UUID(),
                patientId: patientId,
                therapistId: UUID(),
                createdBy: .therapist,
                event: event,
                thought: thought,
                feeling: feeling,
                feelingIntensity: feelingIntensity,
                behaviour: behaviour,
                physicalSymptoms: symptoms,
                createdAt: .now,
                updatedAt: .now
            )
            upsertCache(entry)
            return entry
        }
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        guard let patientUUID = patientId.uuidValue else { throw AuthError.notConfigured }
        let therapistId = try await requireTherapistId()
        do {
            let saved: DiaryOneEntry = try await client.from("diary_one_entries")
                .insert(
                    NewDiaryOneEntry(
                        patientId: patientUUID,
                        therapistId: therapistId,
                        createdBy: .therapist,
                        event: event,
                        thought: thought,
                        feeling: feeling,
                        feelingIntensity: feelingIntensity,
                        behaviour: behaviour,
                        physicalSymptoms: symptoms
                    )
                )
                .select(
                    "id, patient_id, therapist_id, created_by, event, thought, feeling, feeling_intensity, behaviour, physical_symptoms, created_at, updated_at"
                )
                .single()
                .execute()
                .value
            upsertCache(saved)
            AppLog.store.info("Diary 1 entry created")
            return saved
        } catch {
            AppLog.store.error(
                "Diary 1 create failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    func updateEntry(
        id: UUID,
        patientId: DatabaseID,
        event: String,
        thought: String,
        feeling: String,
        feelingIntensity: Int,
        behaviour: String,
        physicalSymptoms: String?
    ) async throws -> DiaryOneEntry {
        let symptoms = Self.nullIfEmpty(physicalSymptoms)
        if DemoData.isDemoID(patientId) {
            guard var entry = cached(for: patientId).first(where: { $0.id == id }) else {
                throw AuthError.notConfigured
            }
            entry.event = event
            entry.thought = thought
            entry.feeling = feeling
            entry.feelingIntensity = feelingIntensity
            entry.behaviour = behaviour
            entry.physicalSymptoms = symptoms
            entry.updatedAt = .now
            upsertCache(entry)
            return entry
        }
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        do {
            let saved: DiaryOneEntry = try await client.from("diary_one_entries")
                .update(
                    DiaryOneEntryClinicalUpdate(
                        event: event,
                        thought: thought,
                        feeling: feeling,
                        feelingIntensity: feelingIntensity,
                        behaviour: behaviour,
                        physicalSymptoms: symptoms,
                        updatedAt: Self.timestampString(from: Date())
                    )
                )
                .eq("id", value: id)
                .select(
                    "id, patient_id, therapist_id, created_by, event, thought, feeling, feeling_intensity, behaviour, physical_symptoms, created_at, updated_at"
                )
                .single()
                .execute()
                .value
            upsertCache(saved)
            AppLog.store.info("Diary 1 entry updated")
            return saved
        } catch {
            AppLog.store.error(
                "Diary 1 update failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    func deleteEntry(id: UUID, patientId: DatabaseID) async throws {
        if DemoData.isDemoID(patientId) {
            var list = cached(for: patientId)
            list.removeAll { $0.id == id }
            storeCache(list, for: patientId)
            return
        }
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        do {
            let deleted: [DeletedDiaryOneRow] = try await client.from("diary_one_entries")
                .delete()
                .eq("id", value: id)
                .select("id")
                .execute()
                .value
            guard !deleted.isEmpty else {
                throw AuthError.notConfigured
            }
            var list = cached(for: patientId)
            list.removeAll { $0.id == id }
            storeCache(list, for: patientId)
            AppLog.store.info("Diary 1 entry deleted")
        } catch {
            AppLog.store.error(
                "Diary 1 delete failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    private func cached(for patientId: DatabaseID) -> [DiaryOneEntry] {
        let key = patientId.queryValue
        let list = DemoData.isDemoID(patientId)
            ? (demoEntriesByPatient[key] ?? [])
            : (entriesByPatient[key] ?? [])
        return list.sorted { $0.createdAt > $1.createdAt }
    }

    private func upsertCache(_ entry: DiaryOneEntry) {
        var list = cached(for: entry.patientId).filter { $0.id != entry.id }
        list.append(entry)
        storeCache(list, for: entry.patientId)
    }

    private func storeCache(_ list: [DiaryOneEntry], for patientId: DatabaseID) {
        let sorted = list.sorted { $0.createdAt > $1.createdAt }
        if DemoData.isDemoID(patientId) {
            demoEntriesByPatient[patientId.queryValue] = sorted
        } else {
            entriesByPatient[patientId.queryValue] = sorted
        }
    }

    private func requireTherapistId() async throws -> UUID {
        try await client.auth.session.user.id
    }

    private static func nullIfEmpty(_ raw: String?) -> String? {
        let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? nil : trimmed
    }

    private static func timestampString(from date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: date)
    }
}
